devlib/target.py

```diff
@@ -517,21 +517,26 @@ def path_kind(path):
             src_excep = HostError
             src_path_kind = host_paths_kind
 
-            dst_mkdir = once(self.makedirs)
+            _dst_mkdir = once(self.makedirs)
             dst_path_join = self.path.join
             dst_paths_kind = target_paths_kind
             dst_remove_file = once(functools.partial(self.remove, as_root=as_root))
         elif action == 'pull':
             src_excep = TargetStableError
             src_path_kind = target_paths_kind
 
-            dst_mkdir = once(functools.partial(os.makedirs, exist_ok=True))
+            _dst_mkdir = once(functools.partial(os.makedirs, exist_ok=True))
             dst_path_join = os.path.join
             dst_paths_kind = host_paths_kind
             dst_remove_file = once(os.remove)
         else:
             raise ValueError('Unknown action "{}"'.format(action))
 
+        def dst_mkdir(path):
+            if path:
+                return _dst_mkdir
```

Comment:
Should this not be `return _dst_mkdir(path)`?

Reply:
```
target = Target(kind="linux", host="10.0.0.7", username="root", password="root", name="hikey960")
target.pull('/root/tmp.json', 'foobar')
```

Notice, though, foobar shouldn't exist. In a such case, `dst_kind == None`

Indeed, this should be _dst_mkdir(path)

Reply:
Here's the complete backtrace of the error:

```
Traceback (most recent call last):
  File "./debug-abs-path.py", line 18, in <module>
    time.sleep(1)
  File "/data/work/lisa/lisa/trace.py", line 5735, in __exit__
    x = self._cm.__exit__(*args, **kwargs)
  File "/usr/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
  File "/data/work/lisa/lisa/trace.py", line 5727, in _make_cm
    yield
  File "/usr/lib/python3.8/contextlib.py", line 525, in __exit__
    raise exc_details[1]
  File "/usr/lib/python3.8/contextlib.py", line 510, in __exit__
    if cb(*exc_details):
  File "/data/work/lisa/lisa/trace.py", line 5723, in __exit__
    return proxy.__exit__(*args, **kwargs)
  File "/data/work/lisa/lisa/trace.py", line 5413, in __exit__
    self.get_data(path)
  File "/data/work/lisa/lisa/trace.py", line 5425, in get_data
    return coll.get_data()
  File "/data/work/lisa/external/devlib/devlib/collector/ftrace.py", line 307, in get_data
    self.target.pull(self.target_output_file, self.output_path, timeout=pull_timeout)
  File "/data/work/lisa/external/devlib/devlib/target.py", line 113, in wrapper
    return f(_self, *args, **kwargs)
  File "/data/work/lisa/external/devlib/devlib/target.py", line 666, in pull
    mapping = self._prepare_xfer('pull', sources, dest, pattern=source if globbing else None, as_root=as_root)
  File "/data/work/lisa/external/devlib/devlib/target.py", line 586, in _prepare_xfer
    return groupby_value({
  File "/data/work/lisa/external/devlib/devlib/target.py", line 587, in <dictcomp>
    src: rewrite_dst(src, dest)
  File "/data/work/lisa/external/devlib/devlib/target.py", line 555, in rewrite_dst
    dst_mkdir(os.path.dirname(dst))
  File "/usr/lib/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
FileNotFoundError: [Errno 2] No such file or directory: ''
```